### ISSUE TITLE
[meta] Consolidate generated ES index templates

### ISSUE THREAD

**Issue by ebeahan** [elastic/ecs#1587]
### Overview

With ECS aligning versioning with the Elastic stack releases starting in 8.0, the ECS tooling could be simplified to generate a single set of Elasticsearch index template artifacts. This single generated version would align with the stack release of the same version.

For example, ECS 8.0 produces sample Elasticsearch index mappings that are compatible with Elastic Stack 8.0.

### Advantages

* Simplify the ECS tooling. No additional code to generate multiple versions of the index templates to support multiple versions of ES or multiple versions of the ES index template APIs.
* Reduce the copies of index templates in the repo; simplified user experience.
* Prevent users from trying to use the curated artifacts with the wrong APIs: [example](https://discuss.elastic.co/t/create-a-custom-ecs-settings-via-github-repo/280769/2?u=ebeahan)
* The sample templates are maintained as examples. They should be used as a starting point for experimentation. Users are still free to generate their own variations of index templates but shouldn't rely on ECS maintaining multiple copies.

### Concerns

* Should the tooling allow users to maintain their own legacy index template artifacts (e.g. for users who maintain customer mappings)?
* The `logstash-output-elasticsearch` plugin [depends](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/Rakefile#L17-L31) on the ECS projects generated, legacy index template artifacts as part of their build for ECS compatibility mode. We will need to remove this dependency before we can remove the generated, legacy templates from ECS.

**Comment by ypid-geberit** [comment]
I was forwarded from #1600 to this issue with my idea to "Support to generate ECS fields as dynamic index template". I would propose to keep discussion of this new idea/feature in #1600. Otherwise, you are welcome to give feedback here.

**Comment by djptek** [comment]
Elasticsearch (7.x) Grok is now ECS compliant see:

https://github.com/elastic/elasticsearch/pull/76885

https://github.com/elastic/elasticsearch/tree/7.x/libs/grok/src/main/resources/patterns/ecs-v1

https://github.com/elastic/elasticsearch/blob/dfc9df8c26dd4bdfd5728c23a2fa35d1b8f86cea/docs/reference/scripting/grok-syntax.asciidoc#migrating-to-elastic-common-schema-ecs

Will there be a dependency to ensure the Templates are aligned with the Grok Patterns?

**Comment by kgeller** [comment]
Removal of ES 6 support for ECS 8: https://github.com/elastic/ecs/pull/1680

**Comment by kgeller** [comment]
ES generated artifacts folder structure update: https://github.com/elastic/ecs/pull/1700
Subsequent Logstash update: https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/1060

**Comment by yaauie** [comment]
After the above restructuring, ECS's [8.0 branch](https://github.com/elastic/ecs/blob/8.0/generated/elasticsearch/legacy/template.json) is still generating [ES7-style legacy templates](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/indices-templates-v1.html) (top-level `"mappings"` and `"settings"` fields), and needs to be updated to generate [ES8-style legacy templates](https://www.elastic.co/guide/en/elasticsearch/reference/8.0/index-templates.html) (top-level `"template"` field with `"mappings"` and `"settings"` sub-fields).

For reference, before the stack-alignment changes, Logstash's Elasticsearch output plugin had been providing ECS 8 support by transforming ES7-style generated templates into ES8-style [during vendoring](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/v11.3.2/Rakefile#L50-L65).

**Comment by yaauie** [comment]
🤦🏼 that is on us. The Elasticsearch output plugin for Logstash [uses the _index_template API when ushing templates to Elasticsearch 8+](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/main/lib/logstash/outputs/elasticsearch/http_client.rb#L411-L413).

Thank you for the clarification. Knowing that the generated artifacts are stable in shape is sufficient for the Logstash team to move forward.

**Comment by ebeahan** [comment]
> The Elasticsearch output plugin for Logstash uses the _index_template API when ushing templates to Elasticsearch 8+.

Thanks for clearing that up. 👍 

@yaauie still unclear about one thing. Does the ES output plugin still need ECS to produce a single v2 index template with ALL ECS fields in one single template? I believe the answer is "yes," but I want to confirm.

**Comment by yaauie** [comment]
No. We can work from the "legacy" v1 templates, and in fact doing so simplifies our codepath. We already translate the "legacy"-style index templates for ECS 1.x when vendoring them for use with Elasticaearch 8, and can use the same code-path with the stack-aligned "legacy"-style templates. Moving the Stack-aligned monolith template to V2 adds complexity because we would need to translate those back into legacy for use with ES 7.x (due to our plugins implementation).

**Comment by kgeller** [comment]
Conclusion:

Since ECS is aligning versioning with the Elastic stack releases starting in 8.0, we have simplified the ECS tooling to generate two sets of Elasticsearch index template artifacts. We will now produce 1) a composable template compatible with the v2 `_index_template` API and 2) a legacy index template compatible with v1 "legacy" `_template` API. 

Previously, at every ECS version, we were generating legacy templates compatible with ES 6 and ES 7, as well as a composable template. We have now consolidated down to, for every ECS version, a single ES 8 compatible legacy template and a single composable template.
